### ISSUE TITLE
Eloquent mutators says TEXT can be casted to array/json

### ISSUE THREAD

**Issue by carestad** [laravel/docs#5001]
As the title says, in the [documentation](https://laravel.com/docs/5.7/eloquent-mutators#array-and-json-casting), it says:

> The array cast type is particularly useful when working with columns that are stored as serialized JSON. For example, if your database has a `JSON` or `TEXT` field type that contains serialized JSON, adding the array cast to that attribute will automatically deserialize the attribute to a PHP array when you access it on your Eloquent model

While, I have not been successful making this work with a `TEXT` column, `LONGTEXT`, however, works. The `JSON` type in MariaDB is also an alias for `LONGTEXT` so I really think the documentation should reflect this somehow. Not sure if this is the case for MySQL as well though.

**Comment by staudenmeir** [comment]
How did the `TEXT` column not work?

The only difference between `TEXT` and `LONGTEXT` is the maximum length.

**Comment by carestad** [comment]
That is what I was thinking as well. Could it be due to MariaDB in 10.2 implementing JSON "support" by aliasing it to `LONGTEXT` and not `TEXT`? My container is running 10.3.11, so it is a quite recent version. Have not tried it in 10.2 or 10.1.

Not sure about the technicalities of why this is the case though. Not sure where in the code the logic for converting DB string data to JSON or PHP arrays is located, and if it is in Laravel or Symfony or Doctirne.

**Comment by staudenmeir** [comment]
What error did you get?

**Comment by carestad** [comment]
I never got any errors, but the data was just not cast when the data was stored in a `TEXT` column, but did after I changed it to `LONGTEXT`, using phpMyAdmin.

So in my model I had

```php
$casts = ['json_data' => 'array']; // also tried «with => 'json'» here
```

And when testing in Tinker for instance:

```php
$mod = MyModel::find(1);
$mod->json_data;
```

Would just output

```
=> "["foo", "bar", "hey"]"
```

instead of the desired output of

```
=> [
     "foo",
     "bar",
     "hey",
   ]
```

Now, _of course_, I can not for the life of me reproduce this behaviour again today. Trying to run migrations that creates both `TEXT`, `JSON`, `LONGTEXT`, and even `VARCHAR` columns seem to cast just fine now. Even fooling around with the collation does not seem to affect it (creating `JSON` columns with migrations seems to set the collation to `utf8mb4_bin`, but changing between that and others also seem to work.)

Closing the issue myself now, and I'll revisit if I am able to reproduce this behaviour properly again. Thank you for your time and help so far though!
